### DISCUSSION TITLE
Dedupe Functionality: Automated Duplicate Data Management for Collections with Comparison and Merge Options

### DISCUSSION THREAD

**Discussion by kajal-g-7span** [directus/directus#23252] [Feature Requests]
### Summary

This feature aims to streamline the management of duplicate data within a collection. Currently, users must manually search for duplicates, compare fields, and delete redundant records. This proposal introduces functionality to automate these tasks by adding a "Find Duplicates" button and options to compare, merge, and delete duplicate entries. The feature will also allow users to set a primary record and merge details from duplicates, with an optional user confirmation step before deletion. By implementing this, data accuracy and management efficiency will be improved, minimizing manual effort and ensuring consistency across collections and third-party integrations.

### Basic Example

Suppose you have a collection named posts. Currently, there is no option to search for duplicate data within this collection. After identifying duplicate entries, you must manually check all the fields and delete the redundant records.

**Basic Example:** If this proposal is implemented, you could use a new API endpoint to automate this process. For instance, you might call an endpoint like `/collections/posts/duplicates` to find duplicates. The response would include duplicate entries, allowing you to review them and perform actions like merging or deleting directly through the API.

### Motivation

We are requesting this feature to address the issue of data duplication, particularly when data has been mistakenly created and integrated with other collections or third-party services. The feature will facilitate the examination of existing data, enable the merging of duplicate entries, update associated references, and ultimately remove redundant data.

**Use Cases:**
- **Data Accuracy:** Improve data accuracy by consolidating duplicates and ensuring that only the most current and complete records are maintained.
- **Efficiency:** Streamline data management processes, reducing manual effort and minimizing errors associated with handling duplicate entries.
- **Integration:** Ensure that data across collections and third-party services remain consistent and up-to-date after merging duplicates.

**Expected Outcome:**
The feature aims to enhance data integrity and management efficiency, providing a reliable way to handle duplicates and maintain accurate, cohesive datasets. If this RFC is not accepted, alternative solutions could involve implementing manual processes or third-party tools to address duplication, although these may lack the integrated benefits and streamlined functionality proposed.

### Detailed Design

**1. Create a Button to Find Duplicate Data, Specific to Each Collection**

- **Overview:**
  A button will be added to the collection management interface, allowing users to initiate a search for duplicate data within a specific collection. This button will be labeled "Find Duplicates" and will be accessible from the collection's settings or main interface.

- **Implementation Details:**
  - **UI/UX:** Add the "Find Duplicates" button to the collection view or settings page. The button should be prominently displayed and accessible to users with appropriate permissions.
  - **Backend Logic:** Upon clicking the button, a query will be executed to identify duplicate records based on a defined set of criteria (e.g., identical fields or a combination of fields). The criteria for duplication can be configured or default to common fields like title, email, or unique identifiers.
  - **Results Display:** Duplicates will be displayed in a table format with options to view details. The table should include columns for key fields and actions to compare or merge entries.

**2. Provide Options to Compare and Merge Duplicate Entries**

- **Overview:**
  After finding duplicate data, users will be able to compare the duplicates side-by-side and merge them into a single entry. This feature ensures data integrity and eliminates redundancy.

- **Implementation Details:**
  - **Comparison View:** 
    - **UI/UX:** Provide a detailed comparison view where users can select duplicate entries to compare. The comparison view will display the fields of the selected records side-by-side for easy evaluation.
    - **Highlight Differences:** Highlight differences between the duplicates to aid in decision-making.
  - **Merge Functionality:**
    - **UI/UX:** Add a "Merge" button in the comparison view. Users can select which entry to keep as the primary and choose which details to merge from other entries.
    - **Backend Logic:** Implement logic to merge data from secondary entries into the primary entry. This involves:
      - **Field Merging:** Allow users to select which fields to merge or overwrite in the primary entry.
      - **Data Integrity:** Ensure that merged data maintains integrity and does not introduce inconsistencies.
      - **Update References:** If the entries are linked to other collections or third-party services, update references accordingly to reflect the merged data.
  - **Confirmation Dialog:** Before finalizing the merge, prompt users with a confirmation dialog to prevent accidental merges.

### Must Have:
- **Duplicate Data Finder:** A tool to identify duplicate entries within the collection.
- **Comparison of Duplicated Data:** Functionality to compare duplicate entries to determine which to keep.
- **Delete Unwanted Duplicates:** Capability to remove redundant entries while retaining one copy.

### Should Have:
- **Set Primary Entry and Merge Details:** Feature to designate one entry as primary and merge details from other duplicates into it.

### Could Have:
- **User Confirmation Before Deletion:** Option to require user confirmation before finalizing the deletion of duplicate entries.

### Won't Have:
- **Keeping Duplicate Data:** The system will not retain duplicate entries; all duplicates will be deleted.

### Drawbacks

Implementing this feature could increase code complexity and maintenance, and complicate user training. It may be better suited for user space to avoid core system impacts and allow flexibility. While it's not a breaking change, migrating existing applications might still require effort. Balancing the benefits with these potential challenges is essential.

### Alternatives

Currently, we run database queries directly to manually find and manage data using traditional methods.

**Alternatives:**
Other designs considered might involve implementing automated tools or interfaces to handle duplicate data more efficiently.

**Impact of Not Implementing This:**
If we do not implement this feature, we will continue relying on manual processes, which could be time-consuming and prone to errors, potentially affecting data integrity and management efficiency.

### Adoption Strategy

If we implement this proposal, existing Directus developers can adopt it without any breaking changes, as it will simply be an addition of a field. A migration can be written for this update, and it will have minimal impact on other projects within the Directus ecosystem.

**Comment by rijkvanzanten** [comment]
What would you use as the shared source of truth for the duplicates? 🤔 The primary key would always be different so that's out of the question. Does the user specify manually what fields should be checked?

**Reply by kajal-g-7span** [reply]
Yes, we have to define the duplication criteria to find the data. For example, we can select title, description, and user_created for the posts collection.

**Reply by kajal-g-7span** [reply]
@rijkvanzanten We have prepared one design for this functionality.

We propose adding a new deduplication feature to our application, which will be accessible through a new button on the left menu. This feature is designed to help users identify and merge duplicate records across selected tables. The following outlines the suggested changes:

**Screen 1: Accessing the Dedupe Functionality**
- **Proposed Button Placement:** A new button will be added to the left menu, alongside existing modules, for accessing the deduplication tool.
- **Table Selection:** Upon clicking this button, users will be shown a list of tables where deduplication can be applied. This list would be dynamically generated based on the tables available in the database.
- **Field Selection:** Users will need to select the fields they wish to use for identifying duplicates. This selection will allow for more targeted deduplication based on user-defined criteria.

**Screen 2: Viewing and Selecting Duplicates**
- **Duplicate Grouping Display:** After selecting a table and fields, the system would display grouped data, highlighting potential duplicates. The grouping will be similar to the results of the following PostgreSQL query:
```SELECT 
    t.*,
    COUNT(*) OVER (PARTITION BY column1, column2) AS duplicate_count
FROM 
    table_name t
ORDER BY 
    duplicate_count DESC, 
    column1, 
    column2;
```
- This query groups items based on the selected fields and sorts them so that duplicates appear at the top.
- **Item Selection:** Users would then have the option to select which items they believe are duplicates. A Merge Items button would be provided for users to proceed with merging the selected items.

**Screen 3: Merging Duplicate Items**
- **Primary Item Display:** Following the selection of items to merge, the system would show the details of the first chosen item as the primary item. This item will be highlighted for clarity.
- **Field Comparison and Selection:** On the right side of the screen, each field of the primary item would be accompanied by a dropdown menu containing values from the other selected items. Users would compare these values and select which one to retain in the merged item.
- This feature would allow users to ensure that the final merged item contains the most accurate and relevant information.
- **Final Merge:** Once all field selections are made, users would confirm the merge operation, resulting in a single updated record that consolidates the information from the duplicate entries.

<img width="615" alt="image" src="https://github.com/user-attachments/assets/9c78b6bb-aee5-44a9-966d-51dfd8c0ac67">
<img width="531" alt="image" src="https://github.com/user-attachments/assets/a082ba5e-762a-4e42-b4a6-4e68d6221116">
<img width="577" alt="image" src="https://github.com/user-attachments/assets/2fd406cb-7ff7-45e4-aede-4ab42d2ac217">
